promql/functions.go

```diff
@@ -422,6 +455,66 @@ func funcMaxOverTime(ev *evaluator, args Expressions) Value {
 	})
 }
 
+// === first(Vector ValueTypeVector) Vector ===
+func funcFirst(ev *evaluator, args Expressions) Value {
+	vec := ev.evalVector(args[0])
+	resultVector := Vector{}
+	if len(vec) > 0 {
+		resultVector = append(resultVector, vec[0])
+	}
+	return resultVector
+}
+
+// === last(Vector ValueTypeVector) Vector ===
+func funcLast(ev *evaluator, args Expressions) Value {
+	vec := ev.evalVector(args[0])
+	resultVector := Vector{}
+	if len(vec) > 0 {
+		resultVector = append(resultVector, vec[len(vec)-1])
+	}
+	return resultVector
+}
+
+// === first_over_time(Matrix ValueTypeMatrix) Vector ===
+func funcFirstOverTime(ev *evaluator, args Expressions) Value {
+	mat := ev.evalMatrix(args[0])
+	resultVector := Vector{}
+
+	for _, el := range mat {
+		if len(el.Points) == 0 {
+			continue
+		}
+
+		if len(el.Points) > 0 {
+			resultVector = append(resultVector, Sample{
+				Metric: dropMetricName(el.Metric),
+				Point:  el.Points[0],
+			})
+		}
+	}
+	return resultVector
+}
+
+// === last_over_time(Matrix ValueTypeMatrix) Vector ===
+func funcLastOverTime(ev *evaluator, args Expressions) Value {
+	mat := ev.evalMatrix(args[0])
```

Comment:
You should probably use aggrOverTime

Reply:
Yes, done.  I think I was thinking about the case when there would be no data returned, but this is not the case.